File: src/models/image_dir.py
from typing import ClassVar, Mapping, Any, Dict, Optional, Tuple, List, Sequence
from typing_extensions import Self

from typing import NamedTuple

from PIL import Image

from viam.media.video import NamedImage
from viam.proto.common import ResponseMetadata

from viam.module.types import Reconfigurable
from viam.proto.app.robot import ComponentConfig
from viam.proto.common import ResourceName
from viam.resource.base import ResourceBase
from viam.resource.types import Model, ModelFamily
from viam.components.component_base import ValueTypes

from viam.components.camera import Camera, ViamImage
from viam.media.utils.pil import pil_to_viam_image, CameraMimeType

from viam.logging import getLogger
from viam.errors import ViamError

import os
from datetime import datetime, timezone

import re

LOGGER = getLogger(__name__)
SUPPORTED_EXTS = {"jpg", "jpeg", "png", "gif"}

class imageDir(Camera, Reconfigurable):
    class Properties(NamedTuple):
        supports_pcd: bool = False
        intrinsic_parameters: Optional[Any] = None
        distortion_parameters: Optional[Any] = None

    MODEL: ClassVar[Model] = Model(ModelFamily("viam-labs", "camera"), "image-dir")

    camera_properties: Camera.Properties = Properties()
    # will store current get_image index for a given directory here
    directory_index: dict
    root_dir: str = "/tmp"
    ext: str = "jpg"
    sub_dir: str = "" 
    
    # Precomputed once per configured (dir, ext)
    sorted_files: List[str] = [] 
    sub_dir_len: int = 0 
    
    # Constructor
    @classmethod
    def new(
        cls, config: ComponentConfig, dependencies: Mapping[ResourceName, ResourceBase]
    ) -> Self:
        my_class = cls(config.name)
        my_class.reconfigure(config, dependencies)
        return my_class
    
    @classmethod
    def validate_config(cls, config: ComponentConfig) -> Tuple[Sequence[str], Sequence[str]]:
        """Validates  the configuration for the imageDir camera."""
        errors: List[str] = []
        warnings: List[str] = []

        attrs = config.attributes.fields
        root_dir = attrs.get("root_dir").string_value if "root_dir" in attrs else "/tmp"
        ext = attrs.get("ext").string_value if "ext" in attrs else "jpg"
        sub_dir = attrs.get("dir").string_value if "dir" in attrs else ""

        # Root dir must exist
        if not os.path.isdir(root_dir):
            errors.append(f"specified 'root_dir' does not exist: {root_dir}")

        # ext must be supported
        if ext and ext.lower() not in SUPPORTED_EXTS:
            errors.append(f"unsupported 'ext': {ext}. Supported: {sorted(SUPPORTED_EXTS)}")

        # dir must be provided
        if not sub_dir:
            errors.append("'dir' is required and must be a subdirectory of 'root_dir'")

        # If prior checks passed, validate the requested directory and files
        if not errors:
            requested_dir = os.path.join(root_dir, sub_dir)
            if not os.path.isdir(requested_dir):
                errors.append(f"requested 'dir' not found within configured 'root_dir': {requested_dir}")
            else: 
                files = [f for f in os.listdir(requested_dir) if f.lower().endswith(f".{ext.lower()}")]
                if not files:
                    errors.append(f"no files ending with .{ext} found in {requested_dir}")

        return errors, warnings
    # Handles attribute reconfiguration
    def reconfigure(
        self, config: ComponentConfig, dependencies: Mapping[ResourceName, ResourceBase]
    ):
        attrs = config.attributes.fields
        self.root_dir = attrs.get("root_dir").string_value if "root_dir" in attrs else "/tmp"
        self.ext = (attrs.get("ext").string_value if "ext" in attrs else "jpg").lower()
        self.sub_dir = attrs.get("dir").string_value if "dir" in attrs else ""

        # Log after attributes are set
        LOGGER.info(
            "image-dir: reconfigured name=%s root=%s sub_dir=%s ext=%s",
            self.name, self.root_dir, self.sub_dir, self.ext
        )

        requested_dir = os.path.join(self.root_dir, self.sub_dir) if self.sub_dir else self.root_dir
        
        # Build the fixed, chronologically sorted image list (files in the directory) once
        self.sorted_files = self._get_sorted_files(requested_dir, self.ext)
        self.sub_dir_len = len(self.sorted_files)
        
        if self.sub_dir_len == 0:
            raise ViamError(f"No images with valid timestamp or numeric index found in {requested_dir}")
        
        self.directory_index = {requested_dir: 0}
    async def get_image(
        self,
        mime_type: str = "image/jpeg",
        *,
        timeout: Optional[float] = None,
        extra: Optional[Mapping[str, Any]] = None,
        metadata: Optional[Mapping[str, Any]] = None,
        **kwargs,
    ) -> ViamImage:
        extra = extra or {}

        # This code prevents changing dir/ext on a per-call basis
        # Enforces reconfigure() for such changes
        if extra.get("dir") and extra["dir"] != self.sub_dir:
            raise ViamError("Per-call 'dir' override not supported; reconfigure instead.")
        if extra.get("ext") and extra["ext"].lower() != self.ext:
            raise ViamError("Per-call 'ext' override not supported; reconfigure instead.")

        requested_dir = os.path.join(self.root_dir, self.sub_dir)
        if not os.path.isdir(requested_dir):
            raise ViamError("configured directory no longer exists")
        if self.sub_dir_len == 0 or not self.sorted_files:
            raise ViamError("No images preloaded. Did reconfigure fail?")

        image_index: int

        # Direct index setting
        if extra.get("index") is not None:
            image_index = extra["index"]
        # Reset to 0
        elif extra.get("index_reset") is not None and extra["index_reset"]:
            image_index = 0
        # Jog from current position
        elif extra.get("index_jog") is not None:
            image_index = self._jog_index(extra["index_jog"], requested_dir)
        # Use current directory index
        elif self.directory_index.get(requested_dir) is not None:
            image_index = self.directory_index[requested_dir]
        # Default: Start at 0
        else:
            image_index = 0

        # Wrap with precomputed length and use precomputed sorted list
        n = self.sub_dir_len
        image_index = int(image_index) % n

        file_path = os.path.join(requested_dir, self.sorted_files[image_index])
        if not os.path.isfile(file_path):
            raise ViamError(f"Image file not found: {file_path}")

        # Resolve mime type
        if isinstance(mime_type, CameraMimeType):
            mt = mime_type
        else:
            mt_str = mime_type.strip().lower() if mime_type is not None else ""
            # treat these as "no preference" and pick by ext
            if mt_str in {"", "-"}:
                mt_str = "image/png" if self.ext.lower() == "png" else "image/jpeg"
            if mt_str == "image/jpg":
                mt_str = "image/jpeg"
            try:
                mt = CameraMimeType.from_string(mt_str)
            except Exception:
                LOGGER.warning("Unsupported mime '%s'; defaulting to image/jpeg", mt_str)
                mt = CameraMimeType.JPEG

        with Image.open(file_path) as img:
            vi_img = pil_to_viam_image(img.convert("RGB"), mt)

        # Increment safely with modulo 
        self.directory_index[requested_dir] = (image_index + 1) % n 

        return vi_img
    

    def _parse_timestamp_from_filename(self, filename: str) -> Optional[datetime]:
        # Only strip a known image extension; otherwise keep the full name
        name_without_ext = filename
        if "." in filename:
            base, ext = filename.rsplit(".", 1)
            if ext.lower() in SUPPORTED_EXTS:
                name_without_ext = base

        pattern = r"^(\d{4})-(\d{2})-(\d{2})T(\d{2})_(\d{2})_(\d{2})(?:\.(\d{3}))?Z"
        match = re.match(pattern, name_without_ext)
        if not match:
            return None

        year, month, day, hour, minute, second, ms = match.groups()
        try:
            return datetime(
                int(year), int(month), int(day),
                int(hour), int(minute), int(second),
                int((ms or "000")) * 1000,
                tzinfo=timezone.utc,
            )
        except ValueError as e:
            LOGGER.warning(f"Failed to parse timestamp from {filename}: {e}")
            return None
        
    def _extract_index(self, filename: str) -> tuple[Optional[str], Optional[int]]:
        """
        Decide an ordering key for a file.
        Returns (kind, key):
        - ('ts', microseconds_since_midnight) if a timestamp is present
        - ('idx', integer_index) if a trailing integer is present
        - (None, None) if neither is present
        """
        # Try timestamp first (re-uses your existing parser)
        dt = self._parse_timestamp_from_filename(filename)
        if dt is not None:
            key = ((dt.hour * 60 + dt.minute) * 60 + dt.second) * 1_000_000 + dt.microsecond
            return 'ts', key

        # Then try trailing integer index
        base = filename.rsplit('.', 1)[0] if '.' in filename else filename
        m = re.search(r'(\d+)$', base)
        if m:
            try:
                return 'idx', int(m.group(1))
            except ValueError:
                pass

        return None, None

    def _get_sorted_files(self, dir, ext):
        files = [f for f in os.listdir(dir) if f.lower().endswith(f".{ext.lower()}")]
        if not files:
            return []

        ts, idx = [], []
        for f in files:
            kind, key = self._extract_index(f)
            if kind == 'ts':
                ts.append((f, key))
            elif kind == 'idx':
                idx.append((f, key))
            # else: unknown for now; we may warn after choosing mode

        # Sort by majority if mixed filenames: tie → timestamp
        if len(ts) >= len(idx) and len(ts) > 0:
            keep = {f for f, _ in ts}
            for f in files:
                if f not in keep:
                    LOGGER.warning(f"Skipping file without parsable timestamp: {f}")
            ts.sort(key=lambda x: (x[1], x[0]))  # stable & deterministic
            return [f for f, _ in ts]

        # Sort by numeric index
        if len(idx) > 0:
            keep = {f for f, _ in idx}
            for f in files:
                if f not in keep:
                    LOGGER.warning(f"Skipping file without numeric index: {f}")
            idx.sort(key=lambda x: (x[1], x[0]))
            return [f for f, _ in idx]

        # Nothing matched either pattern
        for f in files:
            LOGGER.warning(f"Skipping file without timestamp or numeric index: {f}")
        return []

    def _jog_index(self, index_jog, requested_dir):
        n = self.sub_dir_len if self.sub_dir_len > 0 else 1
        current_index = self.directory_index.get(requested_dir, 0) % n
        
        return (current_index + int(index_jog)) % n

    async def get_images(
        self,
        *,
        timeout: Optional[float] = None,
        metadata: Optional[Mapping[str, Any]] = None,
        extra: Optional[Mapping[str, Any]] = None,
        filter_source_names: Optional[List[str]] = None,
        **kwargs,
    ) -> Tuple[List[NamedImage], ResponseMetadata]:
        if extra is None:
            extra = {}

        # Determine source name
        source_name = self.sub_dir or ""

        # Apply filtering if specified
        if filter_source_names is not None and len(filter_source_names) > 0 and source_name not in filter_source_names:
            return [], ResponseMetadata()

        # Get the image
        image = await self.get_image(extra=extra, timeout=timeout)

        # Create NamedImage
        named_image = NamedImage(name=source_name, data=image.data, mime_type=image.mime_type)

        # Return with metadata
        return [named_image], ResponseMetadata()

    async def get_point_cloud(
        self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs
    ) -> Tuple[bytes, str]:
        raise NotImplementedError()

    # Implements the do_command which will respond to a map with key "request"
    async def do_command(
        self, command: Mapping[str, ValueTypes], *, timeout: Optional[float] = None, **kwargs
    ) -> Mapping[str, ValueTypes]:
        ret = {}
        if command.get("set") is not None:
            setDict = command.get("set")
            # Keep image list fixed; require reconfigure() to change dir/ext
            if setDict.get("dir") is not None or setDict.get("ext") is not None:
                raise ViamError("Changing 'dir' or 'ext' requires a reconfigure.")
            
            requested_dir = os.path.join(self.root_dir, self.sub_dir)
            
            if setDict.get("index") is not None:
                n = max(1, self.sub_dir_len)
                self.directory_index[requested_dir] = setDict["index"] % n
                ret = {"index": self.directory_index[requested_dir]}
                
            if setDict.get("index_reset") is not None and setDict["index_reset"]:
                self.directory_index[requested_dir] = 0
                ret = {"index": 0}
            
            if setDict.get("index_jog") is not None:
                idx = self._jog_index(setDict["index_jog"], requested_dir)
                self.directory_index[requested_dir] = idx
                ret = {"index": idx}

        return ret

    async def get_properties(self, *, timeout: Optional[float] = None, **kwargs) -> Properties:
        return self.camera_properties

File: models/imagedir/imagedir.go
// Package imagedir implements a Viam camera that serves images sequentially
// from a directory on the host machine.
package imagedir

import (
	"context"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"sync"

	"go.viam.com/rdk/components/camera"
	"go.viam.com/rdk/data"
	"go.viam.com/rdk/logging"
	"go.viam.com/rdk/pointcloud"
	"go.viam.com/rdk/resource"
	"go.viam.com/rdk/spatialmath"
)

// Model is the resource model this module registers.
var Model = resource.NewModel("viam-labs", "camera", "image-dir")

const (
	defaultRootDir = "/tmp"
	defaultExt     = "jpg"
)

// supportedExts are the file extensions this camera will serve, keyed lowercase.
var supportedExts = map[string]string{
	"jpg":  "image/jpeg",
	"jpeg": "image/jpeg",
	"png":  "image/png",
	"gif":  "image/gif",
}

func sortedSupportedExts() []string {
	out := make([]string, 0, len(supportedExts))
	for e := range supportedExts {
		out = append(out, e)
	}
	sort.Strings(out)
	return out
}

func init() {
	resource.RegisterComponent(camera.API, Model,
		resource.Registration[camera.Camera, *Config]{Constructor: newImageDir})
}

// Config is the JSON configuration for the image-dir camera.
type Config struct {
	// RootDir is the directory that Dir is resolved against. Defaults to /tmp.
	RootDir string `json:"root_dir"`
	// Ext is the image file extension to serve. Defaults to jpg.
	Ext string `json:"ext"`
	// Dir is the subdirectory of RootDir to read images from. Required.
	Dir string `json:"dir"`
}

func (cfg *Config) rootDir() string {
	if cfg.RootDir == "" {
		return defaultRootDir
	}
	return cfg.RootDir
}

func (cfg *Config) ext() string {
	if cfg.Ext == "" {
		return defaultExt
	}
	return strings.ToLower(cfg.Ext)
}

// requestedDir resolves the directory images are read from.
func (cfg *Config) requestedDir() string {
	if cfg.Dir == "" {
		return cfg.rootDir()
	}
	return filepath.Join(cfg.rootDir(), cfg.Dir)
}

// Validate checks the config and reports every problem found, matching the
// Python module's behaviour of accumulating errors rather than failing on the first.
func (cfg *Config) Validate(path string) ([]string, []string, error) {
	var errs []string

	if fi, err := os.Stat(cfg.rootDir()); err != nil || !fi.IsDir() {
		errs = append(errs, fmt.Sprintf("specified 'root_dir' does not exist: %s", cfg.rootDir()))
	}

	if _, ok := supportedExts[cfg.ext()]; !ok {
		errs = append(errs, fmt.Sprintf("unsupported 'ext': %s. Supported: %v",
			cfg.Ext, sortedSupportedExts()))
	}

	if cfg.Dir == "" {
		errs = append(errs, "'dir' is required and must be a subdirectory of 'root_dir'")
	}

	// Only inspect the directory once the values above are known good.
	if len(errs) == 0 {
		dir := cfg.requestedDir()
		if fi, err := os.Stat(dir); err != nil || !fi.IsDir() {
			errs = append(errs,
				fmt.Sprintf("requested 'dir' not found within configured 'root_dir': %s", dir))
		} else {
			matches, err := filesWithExt(dir, cfg.ext())
			if err != nil {
				errs = append(errs, err.Error())
			} else if len(matches) == 0 {
				errs = append(errs, fmt.Sprintf("no files ending with .%s found in %s", cfg.ext(), dir))
			}
		}
	}

	if len(errs) > 0 {
		return nil, nil, errors.New(strings.Join(errs, "; "))
	}
	return nil, nil, nil
}

// ErrClosed is returned by methods called after the resource has been closed.
var ErrClosed = errors.New("image-dir camera is closed")

// imageDir serves images from a fixed, pre-sorted list built at construction.
type imageDir struct {
	resource.Named
	resource.AlwaysRebuild

	logger logging.Logger

	rootDir string
	ext     string
	subDir  string
	mimeTy  string

	mu sync.Mutex
	// sortedFiles is fixed for the lifetime of the resource; changing dir or ext
	// requires a reconfigure, which rebuilds the resource. Guarded by mu so Close
	// can release it without racing an in-flight read.
	sortedFiles []string
	index       int
	closed      bool
}

// Close releases the image list and prevents further use of the camera.
//
// The resource holds no file handles or goroutines — files are opened and closed
// per read — so there is nothing to tear down. What Close does provide is the
// half of the resource.Resource contract that matters here: it is idempotent,
// and subsequent calls fail with ErrClosed rather than serving stale state.
func (c *imageDir) Close(ctx context.Context) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.closed {
		return nil
	}
	c.closed = true
	c.sortedFiles = nil
	c.index = 0

	c.logger.Infof("image-dir: closed %s", c.Name().ShortName())
	return nil
}

func newImageDir(
	ctx context.Context,
	_ resource.Dependencies,
	conf resource.Config,
	logger logging.Logger,
) (camera.Camera, error) {
	cfg, err := resource.NativeConfig[*Config](conf)
	if err != nil {
		return nil, err
	}
	if _, _, err := cfg.Validate(""); err != nil {
		return nil, err
	}

	dir := cfg.requestedDir()
	files, err := getSortedFiles(dir, cfg.ext(), logger)
	if err != nil {
		return nil, err
	}
	if len(files) == 0 {
		return nil, fmt.Errorf("no images with valid timestamp or numeric index found in %s", dir)
	}

	logger.Infof("image-dir: configured name=%s root=%s sub_dir=%s ext=%s images=%d",
		conf.ResourceName().ShortName(), cfg.rootDir(), cfg.Dir, cfg.ext(), len(files))

	return &imageDir{
		Named:       conf.ResourceName().AsNamed(),
		logger:      logger,
		rootDir:     cfg.rootDir(),
		ext:         cfg.ext(),
		subDir:      cfg.Dir,
		mimeTy:      supportedExts[cfg.ext()],
		sortedFiles: files,
	}, nil
}

// Images returns the image at the current index and advances the index.
//
// The extra map supports the same keys the Python module accepted:
//
//	index       int  - serve this absolute index (wraps)
//	index_reset bool - serve index 0
//	index_jog   int  - offset from the current index (wraps, may be negative)
func (c *imageDir) Images(
	ctx context.Context,
	filterSourceNames []string,
	extra map[string]interface{},
) ([]camera.NamedImage, resource.ResponseMetadata, error) {
	sourceName := c.subDir

	if len(filterSourceNames) > 0 && !contains(filterSourceNames, sourceName) {
		return nil, resource.ResponseMetadata{}, nil
	}

	// dir and ext are fixed at construction; reject per-call overrides rather
	// than silently ignoring them.
	if v, ok := extra["dir"]; ok {
		if s, _ := v.(string); s != "" && s != c.subDir {
			return nil, resource.ResponseMetadata{},
				errors.New("per-call 'dir' override not supported; reconfigure instead")
		}
	}
	if v, ok := extra["ext"]; ok {
		if s, _ := v.(string); s != "" && !strings.EqualFold(s, c.ext) {
			return nil, resource.ResponseMetadata{},
				errors.New("per-call 'ext' override not supported; reconfigure instead")
		}
	}

	dir := filepath.Join(c.rootDir, c.subDir)
	if fi, err := os.Stat(dir); err != nil || !fi.IsDir() {
		return nil, resource.ResponseMetadata{}, errors.New("configured directory no longer exists")
	}

	// Resolve the index and snapshot the filename under the lock, then read the
	// file outside it so a slow disk doesn't serialise concurrent callers.
	c.mu.Lock()
	if c.closed {
		c.mu.Unlock()
		return nil, resource.ResponseMetadata{}, ErrClosed
	}
	idx, err := c.resolveIndexLocked(extra)
	if err != nil {
		c.mu.Unlock()
		return nil, resource.ResponseMetadata{}, err
	}
	name, n := c.sortedFiles[idx], len(c.sortedFiles)
	c.mu.Unlock()

	path := filepath.Join(dir, name)
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, resource.ResponseMetadata{}, fmt.Errorf("image file not found: %s: %w", path, err)
	}

	named, err := camera.NamedImageFromBytes(raw, sourceName, c.mimeTy, data.Annotations{})
	if err != nil {
		return nil, resource.ResponseMetadata{}, err
	}

	// Advance only after a successful read, so a bad file is retried rather than
	// silently skipped. A Close racing the read wins: the index stays cleared.
	c.mu.Lock()
	if !c.closed {
		c.index = (idx + 1) % n
	}
	c.mu.Unlock()

	return []camera.NamedImage{named}, resource.ResponseMetadata{}, nil
}

// resolveIndexLocked picks the index to serve, honouring the extra overrides.
// Callers must hold c.mu.
func (c *imageDir) resolveIndexLocked(extra map[string]interface{}) (int, error) {
	n := len(c.sortedFiles)
	if n == 0 {
		return 0, errors.New("no images preloaded")
	}

	switch {
	case extra["index"] != nil:
		v, err := toInt(extra["index"])
		if err != nil {
			return 0, fmt.Errorf("'index': %w", err)
		}
		return wrap(v, n), nil
	case isTrue(extra["index_reset"]):
		return 0, nil
	case extra["index_jog"] != nil:
		v, err := toInt(extra["index_jog"])
		if err != nil {
			return 0, fmt.Errorf("'index_jog': %w", err)
		}
		return wrap(wrap(c.index, n)+v, n), nil
	default:
		return wrap(c.index, n), nil
	}
}

// DoCommand supports {"set": {...}} to move the playback index.
func (c *imageDir) DoCommand(ctx context.Context, cmd map[string]interface{}) (map[string]interface{}, error) {
	ret := map[string]interface{}{}

	rawSet, ok := cmd["set"]
	if !ok || rawSet == nil {
		return ret, nil
	}
	set, ok := rawSet.(map[string]interface{})
	if !ok {
		return nil, errors.New("'set' must be a map")
	}

	if set["dir"] != nil || set["ext"] != nil {
		return nil, errors.New("changing 'dir' or 'ext' requires a reconfigure")
	}

	c.mu.Lock()
	defer c.mu.Unlock()

	if c.closed {
		return nil, ErrClosed
	}

	n := len(c.sortedFiles)
	if n == 0 {
		return nil, errors.New("no images preloaded")
	}

	if v := set["index"]; v != nil {
		i, err := toInt(v)
		if err != nil {
			return nil, fmt.Errorf("'index': %w", err)
		}
		c.index = wrap(i, n)
		ret["index"] = c.index
	}

	if isTrue(set["index_reset"]) {
		c.index = 0
		ret["index"] = 0
	}

	if v := set["index_jog"]; v != nil {
		i, err := toInt(v)
		if err != nil {
			return nil, fmt.Errorf("'index_jog': %w", err)
		}
		c.index = wrap(wrap(c.index, n)+i, n)
		ret["index"] = c.index
	}

	return ret, nil
}

func (c *imageDir) Properties(ctx context.Context) (camera.Properties, error) {
	return camera.Properties{
		SupportsPCD: false,
		MimeTypes:   []string{c.mimeTy},
	}, nil
}

func (c *imageDir) NextPointCloud(ctx context.Context, extra map[string]interface{}) (pointcloud.PointCloud, error) {
	return nil, errors.New("image-dir camera does not support point clouds")
}

func (c *imageDir) Geometries(ctx context.Context, extra map[string]interface{}) ([]spatialmath.Geometry, error) {
	return nil, nil
}

// ---- file ordering -------------------------------------------------------

// tsPattern matches the leading timestamp Viam data-capture puts on filenames,
// e.g. 2026-05-30T11_36_12.018Z_...
var tsPattern = regexp.MustCompile(`^(\d{4})-(\d{2})-(\d{2})T(\d{2})_(\d{2})_(\d{2})(?:\.(\d{3}))?Z`)

// trailingIntPattern matches a trailing integer, e.g. frame_12
var trailingIntPattern = regexp.MustCompile(`(\d+)$`)

func filesWithExt(dir, ext string) ([]string, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}
	suffix := "." + strings.ToLower(ext)
	var out []string
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		if strings.HasSuffix(strings.ToLower(e.Name()), suffix) {
			out = append(out, e.Name())
		}
	}
	return out, nil
}

// stripKnownExt removes the extension only when it is one this module supports,
// so a dot elsewhere in the name is left alone.
func stripKnownExt(filename string) string {
	i := strings.LastIndex(filename, ".")
	if i < 0 {
		return filename
	}
	if _, ok := supportedExts[strings.ToLower(filename[i+1:])]; ok {
		return filename[:i]
	}
	return filename
}

// timeOfDayKey returns microseconds since midnight for a timestamped filename.
//
// NOTE: this deliberately ignores the calendar date, matching the Python
// module it was ported from. Files spanning multiple days therefore interleave
// by time of day rather than sorting chronologically.
func timeOfDayKey(filename string) (int64, bool) {
	m := tsPattern.FindStringSubmatch(stripKnownExt(filename))
	if m == nil {
		return 0, false
	}
	atoi := func(s string) int64 { v, _ := strconv.ParseInt(s, 10, 64); return v }

	month, day := atoi(m[2]), atoi(m[3])
	hour, minute, sec := atoi(m[4]), atoi(m[5]), atoi(m[6])
	if month < 1 || month > 12 || day < 1 || day > 31 ||
		hour > 23 || minute > 59 || sec > 59 {
		return 0, false
	}
	var ms int64
	if m[7] != "" {
		ms = atoi(m[7])
	}
	return ((hour*60+minute)*60+sec)*1_000_000 + ms*1000, true
}

func trailingIndexKey(filename string) (int64, bool) {
	base := filename
	if i := strings.LastIndex(filename, "."); i >= 0 {
		base = filename[:i]
	}
	m := trailingIntPattern.FindStringSubmatch(base)
	if m == nil {
		return 0, false
	}
	v, err := strconv.ParseInt(m[1], 10, 64)
	if err != nil {
		return 0, false
	}
	return v, true
}

type keyed struct {
	name string
	key  int64
}

// getSortedFiles builds the fixed playback order. Files are ordered by embedded
// timestamp when most of them carry one, otherwise by trailing integer index.
func getSortedFiles(dir, ext string, logger logging.Logger) ([]string, error) {
	files, err := filesWithExt(dir, ext)
	if err != nil {
		return nil, err
	}
	if len(files) == 0 {
		return nil, nil
	}

	var ts, idx []keyed
	for _, f := range files {
		if k, ok := timeOfDayKey(f); ok {
			ts = append(ts, keyed{f, k})
			continue
		}
		if k, ok := trailingIndexKey(f); ok {
			idx = append(idx, keyed{f, k})
		}
	}

	sortKeyed := func(in []keyed) []string {
		sort.Slice(in, func(i, j int) bool {
			if in[i].key != in[j].key {
				return in[i].key < in[j].key
			}
			return in[i].name < in[j].name // stable & deterministic
		})
		out := make([]string, len(in))
		for i, k := range in {
			out[i] = k.name
		}
		return out
	}

	warnSkipped := func(kept []keyed, why string) {
		if logger == nil {
			return
		}
		in := make(map[string]bool, len(kept))
		for _, k := range kept {
			in[k.name] = true
		}
		for _, f := range files {
			if !in[f] {
				logger.Warnf("Skipping file without %s: %s", why, f)
			}
		}
	}

	// Tie goes to timestamp ordering, matching the Python module.
	if len(ts) >= len(idx) && len(ts) > 0 {
		warnSkipped(ts, "parsable timestamp")
		return sortKeyed(ts), nil
	}
	if len(idx) > 0 {
		warnSkipped(idx, "numeric index")
		return sortKeyed(idx), nil
	}

	warnSkipped(nil, "timestamp or numeric index")
	return nil, nil
}

// ---- small helpers -------------------------------------------------------

// wrap returns v mod n using Python's floor-modulo semantics, so negative
// offsets wrap around the end of the list rather than going out of range.
func wrap(v, n int) int {
	if n <= 0 {
		return 0
	}
	return ((v % n) + n) % n
}

// toInt accepts the numeric shapes that survive a JSON/protobuf round trip.
func toInt(v interface{}) (int, error) {
	switch t := v.(type) {
	case int:
		return t, nil
	case int32:
		return int(t), nil
	case int64:
		return int(t), nil
	case float32:
		return int(t), nil
	case float64:
		return int(t), nil
	case string:
		i, err := strconv.Atoi(t)
		if err != nil {
			return 0, fmt.Errorf("expected an integer, got %q", t)
		}
		return i, nil
	default:
		return 0, fmt.Errorf("expected an integer, got %T", v)
	}
}

func isTrue(v interface{}) bool {
	b, ok := v.(bool)
	return ok && b
}

func contains(haystack []string, needle string) bool {
	for _, s := range haystack {
		if s == needle {
			return true
		}
	}
	return false
}
